engine/src/flutter/lib/web_ui/lib/src/engine/web_paragraph/paint.dart

```diff
@@ -126,12 +125,27 @@ class TextPaint {
           case StyleElements.shadows:
             paintContext.save();
             for (final ui.Shadow shadow in clusterText.style.shadows!) {
-              painter.fillShadow(clusterText, shadow, layout.isDefaultLtr);
+              painter.fillShadow(
+                clusterText,
+                shadow,
+                // We shape ellipsis with default direction coming from the attaching block
+                // and all the other blocks with the default paragraph direction
+                block is EllipsisBlock
+                    ? block.isLtr
+                    : layout.paragraph.paragraphStyle.textDirection == ui.TextDirection.ltr,
+              );
               painter.paintShadow(canvas, sourceRect, targetRect);
             }
             paintContext.restore();
           case StyleElements.text:
-            painter.fillTextCluster(clusterText, layout.isDefaultLtr);
+            painter.fillTextCluster(
+              clusterText,
+              // We shape ellipsis with default direction coming from the attaching block
+              // and all the other blocks with the default paragraph direction
+              block is EllipsisBlock
+                  ? block.isLtr
+                  : layout.paragraph.paragraphStyle.textDirection == ui.TextDirection.ltr,
```

Comment:
Why are we using paragraph direction to draw all blocks, but we use block direction for ellipsis block? The comment above says "what" we are doing but it doesn't say "why".